backend/layers/processing/dataset_metadata_update.py

```diff
@@ -0,0 +1,288 @@
+        self.datasets_bucket = datasets_bucket
+
+    def update_h5ad(
+        self,
+        h5ad_uri: str,
+        current_dataset_version: DatasetVersion,
+        new_key_prefix: str,
+        new_dataset_version_id: DatasetVersionId,
+        metadata_update: DatasetArtifactMetadataUpdate,
+    ):
+        h5ad_filename = self.download_from_source_uri(
+            source_uri=h5ad_uri,
+            local_path=CorporaConstants.LABELED_H5AD_ARTIFACT_FILENAME,
+        )
+
+        adata = scanpy.read_h5ad(h5ad_filename, backed="r")
+        metadata = current_dataset_version.metadata
+        # maps artifact name for metadata field to DB field name, if different
+        for key, val in metadata_update.as_dict_without_none_values().items():
+            adata.uns[key] = val
+
+            db_field = ARTIFACT_TO_DB_FIELD.get(key) if key in ARTIFACT_TO_DB_FIELD else key
+            setattr(metadata, db_field, val)
+
+        adata.write(h5ad_filename, compression="gzip")
+        self.business_logic.set_dataset_metadata(new_dataset_version_id, metadata)
+
+        self.create_artifact(
+            h5ad_filename,
+            DatasetArtifactType.H5AD,
+            new_key_prefix,
+            new_dataset_version_id,
+            self.artifact_bucket,
+            DatasetStatusKey.H5AD,
+            datasets_bucket=self.datasets_bucket,
+        )
+        os.remove(h5ad_filename)
+        self.update_processing_status(
+            new_dataset_version_id, DatasetStatusKey.VALIDATION, DatasetValidationStatus.VALID
+        )
+        self.update_processing_status(new_dataset_version_id, DatasetStatusKey.H5AD, DatasetConversionStatus.CONVERTED)
+
+    def update_rds(
+        self,
+        rds_uri: str,
+        new_key_prefix: str,
+        new_dataset_version_id: DatasetVersionId,
+        metadata_update: DatasetArtifactMetadataUpdate,
+    ):
+        seurat_filename = self.download_from_source_uri(
+            source_uri=rds_uri,
+            local_path=CorporaConstants.LABELED_H5AD_ARTIFACT_FILENAME,
+        )
+        self.update_processing_status(new_dataset_version_id, DatasetStatusKey.RDS, DatasetConversionStatus.CONVERTING)
+
+        rds_object = base.readRDS(seurat_filename)
+
+        for key, val in metadata_update.as_dict_without_none_values().items():
+            seurat_metadata = seurat.Misc(object=rds_object)
+            if seurat_metadata.rx2[key]:
+                val = val if isinstance(val, list) else [val]
+                seurat_metadata[seurat_metadata.names.index(key)] = StrVector(val)
+
+        base.saveRDS(rds_object, file=seurat_filename)
+
+        self.create_artifact(
+            seurat_filename,
+            DatasetArtifactType.RDS,
+            new_key_prefix,
+            new_dataset_version_id,
+            self.artifact_bucket,
+            DatasetStatusKey.RDS,
+            datasets_bucket=self.datasets_bucket,
+        )
+        os.remove(seurat_filename)
```

Comment:
I'm gathering that this helps in scenarios where multiple datasets are being processed at the same time on the same machine...? 🤔

Reply:
It was an OOM safeguard suggestion from @Bento007

Reply:
OOM? It's removing a file from _disk_...

Reply:
I mean I don't think it's going to do anything fundamentally bad but it seems unnecessary and may just cause the entire operation to take a few seconds longer than otherwise required 🤷‍♂️ 

Unless we have reason to believe we'll have disk space issues. I assume no. Anyways nbd either way

Reply:
Sorry, OOM on the brain. It was a suggestion to save on disk space. I don't know how likely it is that we would have a disk space problem, though I agreed it wouldn't hurt